# Introduction to ConvNets

In [2]:
from keras import layers, models

model = models.Sequential()

model.add(layers.Conv2D(32, (3,3), activation='relu', input_shape=(28, 28, 1)))
model.add(layers.MaxPooling2D((2,2)))
model.add(layers.Conv2D(64, (3,3), activation='relu'))
model.add(layers.MaxPooling2D((2,2)))
model.add(layers.Conv2D(64,(3,3), activation='relu'))

In [3]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 26, 26, 32)        320       
                                                                 
 max_pooling2d (MaxPooling2D  (None, 13, 13, 32)       0         
 )                                                               
                                                                 
 conv2d_1 (Conv2D)           (None, 11, 11, 64)        18496     
                                                                 
 max_pooling2d_1 (MaxPooling  (None, 5, 5, 64)         0         
 2D)                                                             
                                                                 
 conv2d_2 (Conv2D)           (None, 3, 3, 64)          36928     
                                                                 
Total params: 55,744
Trainable params: 55,744
Non-traina

The next step is to feed the last output tensor (of shape (3, 3, 64)) into a densely connected classifier network like those you’re already familiar with: a stack of Dense layers. These classifiers process vectors, which are 1D, whereas the current output is a 3D tensor. First we have to flatten the 3D outputs to 1D, and then add a few Dense layers on top

In [4]:
model.add(layers.Flatten())
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(10, activation='softmax'))

In [5]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 26, 26, 32)        320       
                                                                 
 max_pooling2d (MaxPooling2D  (None, 13, 13, 32)       0         
 )                                                               
                                                                 
 conv2d_1 (Conv2D)           (None, 11, 11, 64)        18496     
                                                                 
 max_pooling2d_1 (MaxPooling  (None, 5, 5, 64)         0         
 2D)                                                             
                                                                 
 conv2d_2 (Conv2D)           (None, 3, 3, 64)          36928     
                                                                 
 flatten (Flatten)           (None, 576)               0

As you can see, the (3, 3, 64) outputs are flattened into vectors of shape (576,) before going through two Dense layers.\
>3x3x64=576

Now, let’s train the convnet on the MNIST digits.

In [14]:
from keras.datasets import mnist
from keras.utils import to_categorical

(train_images, train_labels),(test_images, test_labels) = mnist.load_data()

In [15]:
len(train_images)

60000

In [16]:
len(test_images)

10000

In [17]:
train_images = train_images.reshape((60000, 28, 28, 1))
train_images = train_images.astype('float32') / 255

test_images = test_images.reshape((10000, 28, 28, 1))
test_images = test_images.astype('float32') / 255

In [18]:
train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

#just to see how the first element of labels look like after converting to categorical
train_labels[0] 

array([0., 0., 0., 0., 0., 1., 0., 0., 0., 0.], dtype=float32)

In [19]:
model.compile(optimizer='rmsprop', loss='categorical_crossentropy',
              metrics=['accuracy']
             )

In [20]:
model.fit(train_images,train_labels,
          epochs=5, batch_size=64
         )

Epoch 1/5
938/938 [==============================] - 23s 23ms/step - loss: 0.1777 - accuracy: 0.9440
Epoch 2/5
938/938 [==============================] - 24s 26ms/step - loss: 0.0463 - accuracy: 0.9856
Epoch 3/5
938/938 [==============================] - 25s 27ms/step - loss: 0.0309 - accuracy: 0.9903
Epoch 4/5
938/938 [==============================] - 27s 29ms/step - loss: 0.0239 - accuracy: 0.9926
Epoch 5/5
938/938 [==============================] - 27s 28ms/step - loss: 0.0182 - accuracy: 0.9944


In [21]:
#evaluate the model
test_loss, test_acc = model.evaluate(test_images, test_labels)

test_acc

313/313 [==============================] - 2s 6ms/step - loss: 0.0288 - accuracy: 0.9915


0.9915000200271606

In [22]:
test_loss

0.028779286891222

Whereas the densely connected network from a previous notebook had a test accuracy of 97.8%, the basic convnet has a test accuracy of 99.3%: we decreased the error rate by 68%
(relative). Not bad!